pkg/controllers/nodeclaim/podevents/controller.go

```diff
@@ -0,0 +1,134 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package podevents
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/clock"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	v1 "sigs.k8s.io/karpenter/pkg/apis/v1"
+	nodeutils "sigs.k8s.io/karpenter/pkg/utils/node"
+	podutils "sigs.k8s.io/karpenter/pkg/utils/pod"
+)
+
+// dedupeTimeout is 10 seconds to reduce the number of writes to the APIServer, since pod scheduling and deletion events are very frequent.
+// The smaller this value is, the more writes Karpenter will make in a busy cluster. This timeout is intentionally smaller than the consolidation
+// 15 second validation period, so that we can ensure that we invalidate consolidation commands that are decided while we're de-duping pod events.
+const dedupeTimeout = 10 * time.Second
+
+// Podevents is a nodeclaim controller that deletes adds the lastPodEvent status onto the nodeclaim
+type Controller struct {
+	clock      clock.Clock
+	kubeClient client.Client
+}
+
+// NewController constructs a nodeclaim disruption controller
+func NewController(clk clock.Clock, kubeClient client.Client) *Controller {
+	return &Controller{
+		kubeClient: kubeClient,
+		clock:      clk,
+	}
+}
+
+//nolint:gocyclo
+func (c *Controller) Reconcile(ctx context.Context, pod *corev1.Pod) (reconcile.Result, error) {
+	// If the pod doesn't have a node name, we don't know which node this pod refers to.
+	// or if this is a daemonset
+	if pod.Spec.NodeName == "" || podutils.IsOwnedByDaemonSet(pod) {
+		return reconcile.Result{}, nil
+	}
+
+	node := &corev1.Node{}
+	if err := c.kubeClient.Get(ctx, types.NamespacedName{Name: pod.Spec.NodeName}, node); err != nil {
+		return reconcile.Result{}, client.IgnoreNotFound(fmt.Errorf("getting node, %w", err))
+	}
+
+	// If the node isn't owned by Karpenter, don't do anything
+	if _, ok := node.Labels[v1.NodePoolLabelKey]; !ok {
+		return reconcile.Result{}, nil
+	}
+
+	// If there's no associated node claim, it's not a karpenter owned node.
+	nodeClaims, err := nodeutils.GetNodeClaims(ctx, node, c.kubeClient)
+	if err != nil {
+		return reconcile.Result{}, client.IgnoreNotFound(fmt.Errorf("getting nodeclaims for node, %w", err))
```

Comment:
There's ways to ignore errors within the `nodeutils` file for NodeClaims that I think are different than `client.IgnoreNotFound`. We should use those functions instead because I believe we return custom errors here

Reply:
That's for nodeclaim -> node. We're looking for node first since we have the pod's nodename.